### PR TITLE
Fix memory leaks

### PR DESCRIPTION
**Pull request by lich426**
Which branch your pull-request should merge into?

- `2.1`: BugFixes and Improvements for Current LTS branch

### For each PR
- [ ] Add Copyright if it missed:   
      - `"Copyright (c) 2019-present Axmol Engine contributors (see AUTHORS.md)."`
- [ ] I have performed a self-review of my code.
       
   Optional:
   - [ ] I have checked readme and add important infos to this PR.
   - [ ] I have added/adapted some tests too.
          
### For core/new feature PR
- [ ] I have checked readme and add important infos to this PR.
- [ ] I have added thorough tests.

### CODE REVIEW

**Comment by rh101** [comment]
I don't see a logged issue regarding the memory leaks, and there is no information included in the description of this PR, which would be helpful in order to give some context to the changes made in this PR.

Can you please provide that information?

File: core/base/AutoreleasePool.h [hunk 1]
```diff
@@ -122,6 +122,8 @@ class AX_DLL AutoreleasePool
      */
     void dump();
 
+    void setNeedPop(bool needPop);
```
Comment (rh101) [inline core/base/AutoreleasePool.h:133]: Please explain the purpose of this method, and at least add a comment above it with the relevant info.

**Comment by lich426** [comment]
> I don't see a logged issue regarding the memory leaks, and there is no information included in the description of this PR, which would be helpful in order to give some context to the changes made in this PR.
> 
> Can you please provide that information?

I tested it on Windows 11.
You can check the memory leak by changing the main.cpp file as follows.

```c++
#include "main.h"
#include "AppDelegate.h"
#include "axmol.h"
#ifdef _DEBUG
#    include <stdlib.h>
#    include <crtdbg.h>
#endif

USING_NS_AX;

int WINAPI _tWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPTSTR lpCmdLine, int nCmdShow)
{
#ifdef _DEBUG
    _CrtSetDbgFlag(_CRTDBG_ALLOC_MEM_DF | _CRTDBG_LEAK_CHECK_DF);
#endif

    UNREFERENCED_PARAMETER(hPrevInstance);
    UNREFERENCED_PARAMETER(lpCmdLine);

    // create the application instance
#ifdef USE_WIN32_CONSOLE
    AllocConsole();
    freopen("CONIN$", "r", stdin);
    freopen("CONOUT$", "w", stdout);
    freopen("CONOUT$", "w", stderr);
#endif

    // create the application instance
    AppDelegate app;
    int ret = Application::getInstance()->run();

#ifdef USE_WIN32_CONSOLE
    FreeConsole();
#endif
    return ret;
}
```

**Comment by rh101** [comment]
> You can check the memory leak by changing the main.cpp file as follows.

I can see the output when the program exits, but it doesn't explain your changes.  Can you please provide some information regarding your changes, and whether they resolve some or all of the apparent memory leaks.

**Comment by aismann** [comment]
@lich426 
Please  add Copyright if it missed:
- "Copyright (c) 2019-present Axmol Engine contributors (see AUTHORS.md)."

**Comment by lich426** [comment]
> > You can check the memory leak by changing the main.cpp file as follows.
> 
> I can see the output when the program exits, but it doesn't explain your changes. Can you please provide some information regarding your changes, and whether they resolve some or all of the apparent memory leaks.

The screenshot below is the log that comes out after the program ends. (not english)

- before
<br>

- after
<br>

**Comment by halx99** [comment]
well, the director leak, because it's was reinit after mainloop when purge requested.
we nee re-consider to solve both director and quad command indices leak

**Comment by halx99** [comment]
@rh101  @lich426 The PR #1638 will be more clearly.

**Pull request closed without merging**